File: 1912/Domain/Effects.cs
using System;
using System.Collections.Generic;
using System.Text;


namespace _1912.Domain
{
    class Effects
    {
        public int DiceBonus { get; private set; }
        public int Attack { get; private set; }
        public int AttackKnockout { get; private set; }
        public int AttackCommander { get; private set; }
        public bool AttackBoth { get; private set; }
        public int DefenceCommander { get; private set; }
        public int DefenceKnockout { get; private set; }
        public int DefenceBattalion { get; private set; }
        public int HealCommander { get; private set; }
        public int HealCamp { get; private set; }
        public bool KeepOperation { get; private set; }
        public bool UniqueBattalionAbility { get; private set; }
        public Effects(Card cannon, Card battalion, Card operation, int Dice = -100, int CommanderHealth = 0)
        {
            CheckBattalion(battalion);
            CheckCannon(cannon);
            CheckBoth(cannon, battalion);
            CheckDiceEffects(battalion, operation, Dice, CommanderHealth);
        }

        public Effects()
        {

        }
        private void CheckBoth(Card cannon, Card battalion)
        {
            if (battalion.ID == "battalionWhite" && cannon.ID == "cannonWhite")
            {
                DiceBonus += 1;
            }
            else
            if (battalion.ID == "battalionGreen" && cannon.ID == "cannonGreen")
            {
                Attack += 3;
            }
            else
            if (battalion.ID == "battalionRed" && cannon.ID == "cannonRed")
            {
                Attack += 2;
            }
            else
            if (battalion.ID == "battalionBlue" && cannon.ID == "cannonBlue")
            {
                AttackKnockout += 4;
            }
            else
            if (battalion.ID == "battalionPurple" && cannon.ID == "cannonPurple")
            {
                DefenceBattalion += 2;
            }
        }
        private void CheckCannon(Card cannon)
        {
            if (cannon.ID == "cannonBlue")
            {
                DefenceKnockout += 2;
            }
            if (cannon.ID == "cannonGreen")
            {
                DefenceCommander += 2;
            }
            if (cannon.ID == "cannonSilver")
            {
                AttackBoth = true;
            }
            if (cannon.ID == "cannonWhite")
            {
                DiceBonus += 1;
            }
        }
        private void CheckBattalion(Card battalion)
        {
            if (battalion.ID == "battalionBlue")
            {
                AttackKnockout += 2;
            }
            if (battalion.ID == "battalionPurple")
            {
                AttackCommander += 2;
            }
            if (battalion.ID == "battalionTactical")
            {
                KeepOperation = true;
            }
            if (battalion.ID == "battalionUnique")
            {
                Attack += 2;
            }
            if (battalion.ID == "battalionWhite")
            {
                DiceBonus += 1;
            }
        }
        private void CheckDiceEffects(Card battalion, Card operation, int Dice, int CommanderHealth)
        {
            if (battalion.ID == "battalionGold" && DiceCheck(Dice, new List<int> { 5, 6 }))
            {
                HealCamp += 3;
            }
            if (battalion.ID == "battalionRed" && DiceCheck(Dice, new List<int> { 1, 2 }))
            {
                Attack += 4;
            }
            if (battalion.ID == "cannonLegendary")
            {
                Attack += CommanderHealth;
            }
            if (battalion.ID == "battalionUnique" && DiceCheck(Dice, new List<int> { 4, 5, 6 }))
            {
                UniqueBattalionAbility = true;
            }
            if (battalion.ID == "cannonFamily")
            {
                HealCommander += Dice;
            }
        }
        private bool DiceCheck(int Dice, List<int> values)
        {
            foreach (int i in values)
            {
                if (i <= Dice + DiceBonus && i >= Dice - DiceBonus)
                {
                    return true;
                }
            }
            return false;
        }
    }
}

File: domain/effects.py
from operator import truediv
from domain.card import Card
from typing import List

class Effects:

    dice_bonus = 0
    attack = 0
    attack_knockout = 0
    attack_commander = 0
    attack_both = False
    defence_commander = 0
    defence_knockout = 0
    defence_battalion = 0
    heal_commander = 0
    heal_camp = 0
    keep_operation = False
    unique_battalion_ability = False

    def __init__(self, cannon: Card = None, battalion: Card = None, operation: Card = None, dice: int = -100, commander_health: int = 0):
        if cannon != None:
            self._check_battalion(battalion)
            self._check_cannon(cannon)
            self._check_both(cannon, battalion)
            self._check_dice_effects(battalion, operation, dice, commander_health)
    
    def _check_battalion(self, battalion: Card):
        if battalion.id == "battalionBlue":
            self.attack_knockout += 2
        if battalion.id == "battalionPurple":
            self.attack_commander += 2
        if battalion.id == "battalionTactical":
            self.keep_operation = True
        if battalion.id == "battalionUnique":
            self.attack += 2
        if battalion.id == "battalionWhite":
            self.dice_bonus += 1

    def _check_cannon(self, cannon: Card):
        if cannon.id == "cannonBlue":
            self.defence_knockout += 2
        if cannon.id == "cannonGreen":
            self.defence_commander += 2
        if cannon.id == "cannonSilver":
            self.attack_both = True
        if cannon.id == "cannonWhite":
            self.dice_bonus += 1

    def _check_both(self, cannon: Card, battalion: Card):
        if battalion.id == "battalionWhite" and cannon.id == "cannonWhite":
            self.dice_bonus += 1
        elif battalion.id == "battalionGreen" and cannon.id == "cannonGreen":
            self.attack += 3
        elif battalion.id == "battalionRed" and cannon.id == "cannonRed":
            self.attack += 2
        elif battalion.id == "battalionBlue" and cannon.id == "cannonBlue":
            self.attack_knockout += 4
        elif battalion.id == "battalionPurple" and cannon.id == "cannonPurple":
            self.defence_battalion += 2

    def _check_dice_effects(self, battalion: Card, operation: Card, dice: int, commander_health: int):
        if battalion.id == "battalionGold" and self._dice_check(dice, [5, 6]):
            self.heal_camp += 3
        if battalion.id == "battalionRed" and self._dice_check(dice, [1, 2]):
            self.attack += 4
        if battalion.id == "cannonLegendary":
            self.attack += commander_health
        if battalion.id == "battalionUnique" and self._dice_check(dice, [4, 5, 6]):
            self.unique_battalion_ability = True
        if battalion.id == "cannonFamily":
            self.heal_commander += dice

    def _dice_check(self, dice: int, values: List[int]):
        for v in values:
            if dice - self.dice_bonus <= v <= dice + self.dice_bonus:
                return True
        return False
